### PR TITLE
✨delete provider webhook component with a clusterctl provider delete 

### LINKED ISSUE #2401: clusterctl delete doesn't delete provider webhook component
**What steps did you take and what happened:**
1. Use `clusterctl init --infrastructure aws` to install provider components
1. Use `clusterctl delete --infrastructure aws`.
 I noticed `capi-webhook-system/capa-controller-manager` is still around.

**What did you expect to happen:**
I expected it to get deleted.

**Anything else you would like to add:**
If I do `clusterctl delete --infrastructure aws --include-crd` then, the `capi-webhook-system/capa-controller-manager` get deleted.

I'm not sure what the expected behavior is.

**Environment:**

- Cluster-api version: 1435eed4f1f4f68458be3bb22e03345c19e6fa01

/kind bug
/area clusterctl

### PR DESCRIPTION
**Pull request by nader-ziada**
**What this PR does / why we need it**:
- include the provider webhook component  when deleting an provider with `clusterctl delete` 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2401

### CODE REVIEW

File: cmd/clusterctl/pkg/client/cluster/components.go [hunk 1]
```diff
@@ -116,9 +116,7 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
 	}
 
 	namespaces := []string{options.Provider.Namespace}
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/components.go]:
```suggestion
	namespaces := []string{options.Provider.Namespace, repository.WebhookNamespaceName}
```

File: cmd/clusterctl/pkg/client/cluster/components.go [hunk 2]
```diff
@@ -130,13 +128,25 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
...
-		// NB. Skipping CRDs deletion ensures that also the objects of Kind defined in the CRDs Kind are not deleted.
-		isSharedResource := util.IsSharedResource(obj)
-		if !options.IncludeCRDs && isSharedResource {
-			continue
+
+		isWebhookForProvider := false
+		if obj.GetNamespace() == repository.WebhookNamespaceName {
+			providerLabelName, ok := obj.GetLabels()[clusterv1.ProviderLabelName]
+			if ok {
+				if providerLabelName == options.Provider.ManifestLabel() {
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/components.go]:
```suggestion
			if ok && providerLabelName == options.Provider.ManifestLabel() {
```

File: cmd/clusterctl/pkg/client/cluster/components_test.go [hunk 3]
```diff
@@ -171,7 +171,7 @@ func Test_providerComponents_Delete(t *testing.T) {
 				{object: corev1.ObjectReference{APIVersion: "apiextensions.k8s.io/v1beta1", Kind: "CustomResourceDefinition", Name: "crd1"}, deleted: false},             // crd should be preserved
 				{object: corev1.ObjectReference{APIVersion: "admissionregistration.k8s.io/v1beta1", Kind: "MutatingWebhookConfiguration", Name: "mwh1"}, deleted: false}, // MutatingWebhookConfiguration should be preserved
 				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Namespace", Name: repository.WebhookNamespaceName}, deleted: false},                             // capi-webhook-system namespace should never be deleted
-				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: repository.WebhookNamespaceName, Name: "podx"}, deleted: false},                // provider objects in the capi-webhook-system namespace should be preserved
+				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: repository.WebhookNamespaceName, Name: "podx"}, deleted: true},                 // provider objects in the capi-webhook-system namespace should be preserved
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/components_test.go]: Update comment
```suggestion
				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: repository.WebhookNamespaceName, Name: "podx"}, deleted: true},                 // provider objects in the capi-webhook-system namespace should not be preserved
```

File: cmd/clusterctl/pkg/client/cluster/components_test.go [hunk 4]
```diff
@@ -192,7 +192,7 @@ func Test_providerComponents_Delete(t *testing.T) {
...
```
Comment (wfernandes) [inline cmd/clusterctl/pkg/client/cluster/components_test.go]:
```suggestion
				{object: corev1.ObjectReference{APIVersion: "v1", Kind: "Pod", Namespace: repository.WebhookNamespaceName, Name: "podx"}, deleted: true},                 // provider objects in the capi-webhook-system namespace should not be preserved
```

Comment (nader-ziada) [reply cmd/clusterctl/pkg/client/cluster/components_test.go]: fixed, good catch

File: cmd/clusterctl/pkg/client/cluster/components.go (hunk 2 above)
Comment (nader-ziada) [reply cmd/clusterctl/pkg/client/cluster/components.go]: fixed

**Comment by wfernandes** [comment]
I did some manual testing and it all looks good to me :) 
/lgtm

**Comment by vincepri** [comment]
@nader-ziada Is this correct: we'll only delete the webhook system deployment if the user has asked to delete the CRDs

Code change looks good to me, just want to do double check my assumption

**Comment by nader-ziada** [comment]
@vincepri the original code was only  deleting the webhook component if the user asks to include-crds. This makes makes the it such that the webhook component gets deleted with the provider

**Comment by wfernandes** [comment]
/lgtm cancel
@nader-ziada After re-testing with multiple providers installed, I noticed that this PR deletes the corresponding webhook controller immediately.

As per Vince's comment in the issue:
> This only applies if you're deleting the last of your providers though, for example if you have 2 CAPA installed, we don't want to remove the webhook deployment until the both are deleted.

We should either delete the webhook controller when deleting the last provider or stick to what @fabriziopandini said since documentation says delete it when specifying `--include-crd`.

**Comment by nader-ziada** [comment]
I guess I’ll wait till we decide what the correct behavior is before making more changes

**Comment by nader-ziada** [comment]
after discussing this issue, we decided that current behaviour is correct and there is no need for this change. Closing the PR.

**Pull request closed without merging**